crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -0,0 +1,285 @@
+        let user = get_secret_or_param(&self.params, &self.secret, "auth_user_key", "auth_user");
+        let pass = get_secret_or_param(&self.params, &self.secret, "auth_pass_key", "auth_pass");
+
+        let query = self
+            .params
+            .get("query")
+            .ok_or("Query not found in params".into())
+            .context(super::InvalidConfigurationSnafu {
+                dataconnector: "GraphQL",
+                message: "Query not found",
+            })?
+            .to_owned();
+        let endpoint = Url::parse(&dataset.path()).map_err(Into::into).context(
+            super::InvalidConfigurationSnafu {
+                dataconnector: "GraphQL",
+                message: "Invalid URL",
+            },
+        )?;
+        let json_path = self
+            .params
+            .get("json_path")
+            .map_or("data".to_string(), ToOwned::to_owned);
+
+        let mut headers = HeaderMap::new();
+        headers.append(USER_AGENT, HeaderValue::from_static("spice"));
+        headers.append(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+
+        let mut auth = None;
+        if let Some(token) = token {
+            auth = Some(Auth::Bearer(token));
+        }
+        if let Some(user) = user {
+            auth = Some(Auth::Basic(user, pass));
+        }
+
+        client_builder = client_builder.default_headers(headers);
+
+        Ok(GraphQLClient {
+            client: client_builder.build().map_err(|e| {
+                super::DataConnectorError::InvalidConfiguration {
+                    dataconnector: "GraphQL".to_string(),
+                    message: "Failed to set token".to_string(),
+                    source: e.into(),
+                }
+            })?,
+            query,
+            endpoint,
+            json_path,
+            auth,
+        })
+    }
+}
+
+struct GraphQLTableProvider {
+    client: GraphQLClient,
+    schema: SchemaRef,
+}
+
+impl GraphQLTableProvider {
+    pub async fn new(client: GraphQLClient) -> super::DataConnectorResult<Self> {
+        let (_, schema) = client.execute().await.map_err(Into::into).context(
+            super::UnableToHandleRequestSnafu {
+                dataconnector: "GraphQL",
+            },
+        )?;
+
+        Ok(Self { client, schema })
+    }
+}
+
+#[async_trait]
+impl TableProvider for GraphQLTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> Arc<Schema> {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        let (res, schema) = self.client.execute().await.map_err(to_execution_error)?;
```

Comment:
graphQL should support filter and limit push down?

Reply:
https://github.com/spiceai/spiceai/issues/1593